File: Problem73_PalindromePartitioning.java
import java.util.*;
class Solution {
    List<List<String>> result;
    public List<List<String>> partition(String s) {
        result = new ArrayList<>();
        helper(s, 0, new ArrayList<String>());
        return result;
    }
    
    private void helper(String s, int index, List<String> path) {
        // base case
        if(index >= s.length()) {
            result.add(new ArrayList<>(path));
            return;
        }
        // logic
        for(int i = index; i < s.length(); i++) {
            if(isPalindrome(s, index, i)) {
                path.add(s.substring(index, i + 1));
                helper(s, i + 1, path);   
                // backtrack
                path.remove(path.size() - 1);
            }
        }
    }
    
    private boolean isPalindrome(String s, int left, int right) {
        while(left < right) {
            if(s.charAt(left) != s.charAt(right))
                return false;
            left++;
            right--;
        }
        return true;
    }
}

File: Problem-73 Palindrome Partitioning.py
# 131. Palindrome Partitioning
# https://leetcode.com/problems/palindrome-partitioning/

# Time Complexiety: O(n*(2^n))
# Space Complexiety: O(n)

class Solution:
    def __init__(self):
        self.result = list()
    
    def _helper(self, s, idx, path):
        if idx == len(s):
            self.result.append(path[:])
            return
        
        for i in range(idx, len(s)):
            cur = s[idx:i+1]
            
            if self.isPalindrome(cur):
                path.append(cur)
                self._helper(s,i+1,path)
                path.pop()
    
    def isPalindrome(self, s):
        start = 0
        end = len(s) - 1
        
        while start < end:
            if s[start] != s[end]:
                return False
            start += 1
            end -= 1
        return True
    
    def partition(self, s: str):
        self._helper(s, 0, [])
        return self.result

obj = Solution()
print(obj.partition("aab"))
print(obj.partition("abba"))